main.go

```diff
@@ -202,6 +202,107 @@ func populateList() {
 	}
 }
 
+func filterStringSlice(ss []string, fv string) []string {
+    var rs []string
+    for _, s := range ss {
+        if strings.HasPrefix(s, fv) {
+            rs = append(rs, s)
+        }
+    }
+    return rs
+}
+
+func longestCommonPrefix(ss []string) string {
+    // cover the case where the slice has no or one members
+    switch len(ss) {
+        case 0:
+            return ""
+        case 1:
+            return ss[0]
+    }
+    // all strings are compared by bytes here forward (TBD unicode normalization?)
+    // establish min, max lenth members of the slice by iterating over the members
+    min, max := ss[0], ss[0]
+    for _, s := range ss[1:] {
+        switch {
+            case s < min:
+                min = s
+            case s > max:
+                max = s
+        }
+    }
+    // then iterate over the characters from min to max, as soon as chars don't match return
+    for i := 0; i < len(min) && i < len(max); i++ {
+        if min[i] != max[i] {
+            return min[:i]
+        }
+    }
+    // to cover the case where all members are equal, just return one
+    return min
+}
+
+func stringRemainder(aStr, bStr string) string {
+    var long, short string
+    //figure out which string is longer
+    switch {
+    case len(aStr) < len (bStr):
+        short = aStr
+        long = bStr
+    default:
+        short = bStr
+        long = aStr
+    }
+    // iterate over the strings using an external iterator so we don't lose the value
+    i := 0
+    for i < len(short) && i < len(long) {
+        if short[i] != long[i] {
+            // the strings aren't equal so don't return anything
+            return ""
+        }
+    i++
+    }
+    // return whatever's left of the longer string
+    return long[i:]
+}
+
+func generateTabCompletionSlice(inputWord string) []string {
+    // create a slice to hold the values
+    var firstSlice []string
+    // iterate over all the conversation results
+    for _, s := range channels {
+        if s.MembersType == keybase.TEAM {
+            // its a team so add the topic name as a possible tab completion
+            firstSlice = append(firstSlice, s.TopicName)
+            firstSlice = append(firstSlice, s.Name)
+        } else {
+            // its a user, so clean the name and append the users name as a possible tab completion
+            firstSlice = append(firstSlice, cleanChannelName(s.Name))
+        }
+    }
```

Comment:
What happens when you start typing something like "gen"

Reply:
tab completion will fill `general`
![image](https://user-images.githubusercontent.com/8960497/66587521-cd3c5780-eb47-11e9-99d9-29b6fed0017d.png)
![image](https://user-images.githubusercontent.com/8960497/66587538-d4636580-eb47-11e9-888b-1b5e873ad0d3.png)

Reply:
![tabcomplete](https://user-images.githubusercontent.com/8960497/66588305-691a9300-eb49-11e9-9a7b-a52599c14371.gif)

Reply:
It might be better to look into having it so that it ignores channel when team isn't set? Otherwise should someone make a team/user @general there might be some problems lol Or if there are other channels that are mapped to teams/usernames IE there is an @rudi as well as I am in a team with a channel #rudi

Reply:
ahhh but you missed the reest of the bits.  The flow is this:

1. grab the last word of the input string
2. generate a list of all possible tab completion candidates (even if they are duplicates between teams, channels, and people) that are prefixed by the inputWord
3. determine the longest common prefix (lcp) of all the possible tab completion cantidates
4. insert the longest possible prefix string into the `inputView`

so if your input word is "ru" and your cantidates are `[ "rudi", "rudi9719", and "rudimentary" ]` you would tab-complete only the letters 'di' to make "rudi" and then have to continue typing.

you can see in the gif above i typed "key" and the cantidates were `["keybasefriends", "keybase-pages"]` so initially i got "keybase"  then by typing `f+TAB` i got "keybasefriends"

in short, it behaves exactly the same way that bash completion does!

Reply:
So what happens if you were to simply hit "return" when it filled `/j gen` to `/j general`? Would it prompt for more info or try to join `@general`

Reply:
tab completion is only as smart as the user ;-)

Reply:
That can be said for a lot of things :P Please modify your PR so that channels are only suggested when a team is already present, to avoid unnecessary errors or messages in the wrong holes. A possible (though I wouldn't hold this against the merge, it's fine to not implement this) idea would also be for channels that are unique (#sec_sector_stigis) maybe those could auto-populate themselves entirely _with_ team name? If not, no big deal